File: Kapitel05/ollama-clients/python/demos/streaming.py
#!/usr/bin/env python3
"""
Demo: Streaming mit Ollama

Zeigt die Streaming-API für Echtzeit-Ausgabe.
"""

import ollama
import time


def streaming_chat():
    """Chat mit Streaming-Output."""
    print("=== Streaming Chat Demo ===\n")
    print("Frage: Erkläre den Unterschied zwischen let und const in JavaScript.\n")
    print("Antwort (gestreamt):")

    stream = ollama.chat(
        model="llama3.2",
        messages=[{
            "role": "user",
            "content": "Erkläre den Unterschied zwischen let und const in JavaScript. Maximal 3 Sätze."
        }],
        stream=True
    )

    token_count = 0
    start_time = time.time()

    for chunk in stream:
        content = chunk["message"]["content"]
        print(content, end="", flush=True)
        token_count += 1

    duration = time.time() - start_time
    print(f"\n\nStatistiken:")
    print(f"  - Chunks empfangen: {token_count}")
    print(f"  - Dauer: {duration:.2f}s")
    print(f"  - Geschwindigkeit: ~{token_count / duration:.1f} chunks/s")


def streaming_generate():
    """Generate mit Streaming-Output."""
    print("\n=== Streaming Generate Demo ===\n")
    print("Prompt: Die wichtigsten Programmiersprachen für 2025 sind:\n")

    stream = ollama.generate(
        model="llama3.2",
        prompt="Die wichtigsten Programmiersprachen für 2025 sind:",
        stream=True,
        options={"num_predict": 100}
    )

    for chunk in stream:
        print(chunk["response"], end="", flush=True)

    print("\n")


def streaming_with_timeout():
    """Streaming mit Timeout."""
    print("=== Streaming mit Timeout Demo ===\n")
    print("Stoppe nach 5 Sekunden...\n")

    stream = ollama.chat(
        model="llama3.2",
        messages=[{
            "role": "user",
            "content": "Schreibe eine lange Geschichte über einen Programmierer."
        }],
        stream=True
    )

    start_time = time.time()
    timeout = 5.0

    try:
        for chunk in stream:
            if time.time() - start_time > timeout:
                print("\n\n[Abgebrochen nach 5 Sekunden]")
                break
            print(chunk["message"]["content"], end="", flush=True)
    except KeyboardInterrupt:
        print("\n\n[Manuell abgebrochen]")


def main():
    streaming_chat()
    streaming_generate()
    streaming_with_timeout()


if __name__ == "__main__":
    main()

File: Kapitel05/ollama-clients/typescript/src/demos/streaming.ts
/**
 * Demo: Streaming mit Ollama
 *
 * Zeigt die Streaming-API für Echtzeit-Ausgabe.
 */

import ollama from "ollama";

async function streamingChat() {
  console.log("=== Streaming Chat Demo ===\n");
  console.log("Frage: Erkläre den Unterschied zwischen let und const in JavaScript.\n");
  console.log("Antwort (gestreamt):");

  const response = await ollama.chat({
    model: "llama3.2",
    messages: [
      {
        role: "user",
        content: "Erkläre den Unterschied zwischen let und const in JavaScript. Maximal 3 Sätze.",
      },
    ],
    stream: true,
  });

  let tokenCount = 0;
  const startTime = Date.now();

  for await (const part of response) {
    process.stdout.write(part.message.content);
    tokenCount++;
  }

  const duration = (Date.now() - startTime) / 1000;
  console.log(`\n\nStatistiken:`);
  console.log(`  - Chunks empfangen: ${tokenCount}`);
  console.log(`  - Dauer: ${duration.toFixed(2)}s`);
  console.log(`  - Geschwindigkeit: ~${(tokenCount / duration).toFixed(1)} chunks/s`);
}

async function streamingGenerate() {
  console.log("\n=== Streaming Generate Demo ===\n");
  console.log("Prompt: Die wichtigsten Programmiersprachen für 2025 sind:\n");

  const response = await ollama.generate({
    model: "llama3.2",
    prompt: "Die wichtigsten Programmiersprachen für 2025 sind:",
    stream: true,
    options: {
      num_predict: 100,
    },
  });

  for await (const part of response) {
    process.stdout.write(part.response);
  }

  console.log("\n");
}

async function streamWithAbort() {
  console.log("=== Streaming mit Abbruch Demo ===\n");
  console.log("Stoppe nach 5 Sekunden...\n");

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 5000);

  try {
    const response = await ollama.chat({
      model: "llama3.2",
      messages: [
        {
          role: "user",
          content: "Schreibe eine lange Geschichte über einen Programmierer.",
        },
      ],
      stream: true,
    });

    for await (const part of response) {
      if (controller.signal.aborted) {
        console.log("\n\n[Abgebrochen nach 5 Sekunden]");
        break;
      }
      process.stdout.write(part.message.content);
    }
  } catch (error) {
    if ((error as Error).name === "AbortError") {
      console.log("\n\n[Request abgebrochen]");
    } else {
      throw error;
    }
  } finally {
    clearTimeout(timeout);
  }
}

async function main() {
  await streamingChat();
  await streamingGenerate();
  await streamWithAbort();
}

main().catch(console.error);
